sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/AdaptingIntegerArray.java

```diff
@@ -0,0 +1,193 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.sdk.metrics.internal.state;
+
+import java.util.Arrays;
+import javax.annotation.Nullable;
+
+/**
+ * An integer array that automatically expands its memory consumption (via copy/allocation) when
+ * reaching limits. This assumes counts remain low, to lower memory overhead.
+ *
+ * <p>This class is NOT thread-safe. It is expected to be behind a synchronized incrementer.
+ *
+ * <p>Instances start by attempting to store one-byte per-cell in the integer array. As values grow,
+ * this will automatically instantiate the next-size integer array (byte => short => int => long)
+ * and copy over values into the larger array. This class expects most usage to remain within the
```

Comment:
Is this a valid expectation? If collection interval is 1s, then that's still only 128 recordings per bucket. 128 qps is quite easy on a single app and not all of it will be in the same bucket, it seems reasonable to expect 128 to make it in. 32K qps is much less common though so if sticking with type-switching, I would probably remove the BYTE case and start with SHORT. Though I'm suspicious of the general real world performance benefit of type-switching even if synthetic benchmarks can find cases where it helps.

Reply:
Generally, a lot of latency-based monitoring matches a normal distribution (or a sum of a few normals).  Additionally, if you look at our actual client implementation, we tend to have two types of endpoints:  Those infrequently used (e.g. periodic maintenance type RPCs, metadata things, infrequent user operations, etc.) + those with heavy traffic (core use case).

We need to optimise across this board, and we're creating a LOT of distributions, as we track them by HTTP endpoint right now in instrumentation.   So even if a server has 128 qps, we need endpoints to also see that *and* we need to optimise for high+low traffic scenarios with our defaults.

I do think it's worthwhile to put into place "detection" for these high-qps latency distributions where we can:

- Preserve the previous Scale (assuming we'll be seeing similar traffic in the next collection cycle.
- Preserve the previous sized integer (I was going to work on that as a future optimisation).


For the purposes of this PR though, we're already seeing better performance than MapCounter and I think we have room to go further.

Reply:
Not sure if this changes the conversation, but the default collection interval is 60s.

Reply:
Not really.

If we assume normal distribution (A simple HTTP server usually looks like this.  In practice, you have sum-of-normal distributions when you look at latency for various paths your request will take that alter the normal distribution).

We have 320 buckets to represent counts in boundaries within this normal distribution.   We know about 68% of counts will occur within the stdeviation from the mean.   It's hard to identify how good a job of "tuning" the exponential algorithm can do with aligning its bucket thresholds + scale against the std-deviation, but let's assume we're halfway decent, meaning 160 buckets are within this boundary.

32K qps means we'll have 1,920,000 measurements in a minute = 32K qps * 60s.   If we divide by 160 buckets, that averages about 12,000 per-bucket (naively).  Back-of-the-envelope (we could use normal distribution function to fix this), I'd argue we would probably see peaks into 20k per-bucket.  This is still below the 32k we get from just short alone, and that's for something with relatively high QPS.

Reversing this, let's say if we average 100 counts-per-bucket to fit within byte boundary, I'd say we can measure upwards of 400/500 qps w/ the byte arary.

Remember that, by semantic convention, we're measuring *URL Paths* with individual timeseries, not just a single web endpoint.   This means, e.g. in a REST application we need to allocate a histogram for every single possible endpoint (including things like `GET /entity/{id}`).  It's very likely the <500 qps optimisation in memory usage pays off.   It's entirely likely we never need to upgrade to `long` for real world usage.

This is all back-of-the-envelope math.  I'm sure someone else could do a better calculation for a more realistic example.

Reply:
Makes sense. If we follow through with the optimization to preserve the previously size integer I'd expect it to be nicely adapted to a variety of use cases.

Reply:
I added that optimisation in this PR.